File: 0328. Odd Even Linked List/list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == nullptr)
            return nullptr;
        ListNode *dummy = new ListNode(-1), *dummyTail = dummy, *p = head, *last = p;
        while(p && p->next){
            // pop even node
            ListNode* tmp = p->next;
            last = p;
            p->next = p->next->next;
            p = p->next;
            // insert this node to dummy
            dummyTail->next = tmp;
            dummyTail = tmp;
            tmp->next = nullptr;
        }
        
        // for odd case
        if(last->next)
            last = last->next;
        
        // join the two seperate list
        last->next = dummy->next;
        delete dummy;
        
        return head;
    }
};

File: offer/JZ76 删除链表中重复的结点/list.java
/*
 public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}
*/
public class Solution {
    public ListNode deleteDuplication(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy, cur = head;
        while(cur != null){
            if(cur.next != null && cur.val == cur.next.val){
                // delete
                int val = cur.val;
                while(cur != null){
                    if(cur.val != val)
                        break;
                    cur = cur.next;
                }
                prev.next = cur;
//                 prev = cur;
//                 cur = prev.next;
            }else{
                prev = cur;
                cur = cur.next;
            }
        }
        return dummy.next;
    }
}